iroh-bytes/src/store/bao_file.rs

```diff
@@ -0,0 +1,1238 @@
+                // already in file mode, just write the batch
+                file.write_batch(size, batch)?;
+                Ok(HandleChange::None)
+            }
+            BaoFileStorage::Complete(_) => {
+                // we are complete, so just ignore the write
+                // unless there is a bug, this would just write the exact same data
+                Ok(HandleChange::None)
+            }
+        }
+    }
+}
+
+/// This is finally the thing for which we can implement BaoPairMut.
+///
+/// It is a BaoFileHandle wrapped in an Option, so that we can take it out
+/// in the future.
+#[derive(Debug)]
+pub struct BaoFileWriter(Option<BaoFileHandle>);
+
+impl BaoBatchWriter for BaoFileWriter {
+    async fn write_batch(&mut self, size: u64, batch: Vec<BaoContentItem>) -> std::io::Result<()> {
+        let Some(handle) = self.0.take() else {
+            return Err(io::Error::new(io::ErrorKind::Other, "deferred batch busy"));
+        };
+        let (handle, change) = tokio::task::spawn_blocking(move || {
+            let change = handle.write_batch(size, &batch);
+            (handle, change)
+        })
+        .await
+        .expect("spawn_blocking failed");
+        match change? {
+            HandleChange::None => {}
+            HandleChange::MemToFile => {
+                if let Some(cb) = handle.config.on_file_create.as_ref() {
+                    cb(&handle.hash)?;
+                }
+            }
+        }
+        self.0 = Some(handle);
+        Ok(())
+    }
+
+    async fn sync(&mut self) -> io::Result<()> {
+        let Some(handle) = self.0.take() else {
+            return Err(io::Error::new(io::ErrorKind::Other, "deferred batch busy"));
+        };
+        let (handle, res) = tokio::task::spawn_blocking(move || {
+            let res = handle.storage.write().unwrap().sync_all();
+            (handle, res)
+        })
+        .await
+        .expect("spawn_blocking failed");
+        self.0 = Some(handle);
+        res
+    }
+}
+
+/// A pre order outboard without length prefix.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PreOrderOutboard<R> {
+    /// Root hash
+    pub root: blake3::Hash,
+    /// Tree describing the geometry (size, block size) of the data.
+    pub tree: BaoTree,
+    /// Outboard hash pairs (without length prefix)
+    pub data: R,
+}
+
+impl<R: AsyncSliceReader> Outboard for PreOrderOutboard<R> {
+    fn root(&self) -> blake3::Hash {
+        self.root
+    }
+
+    fn tree(&self) -> BaoTree {
+        self.tree
+    }
+
+    async fn load(&mut self, node: TreeNode) -> io::Result<Option<(blake3::Hash, blake3::Hash)>> {
+        let Some(offset) = self.tree.pre_order_offset(node) else {
+            return Ok(None);
+        };
+        let offset = offset * 64;
+        let content = self.data.read_at(offset, 64).await?;
+        Ok(Some(if content.len() != 64 {
+            (blake3::Hash::from([0; 32]), blake3::Hash::from([0; 32]))
```

Comment:
why is this returning an "empty" hash when the content length is not a multiple of 64?

Reply:
If the file is partial, and you read beyond the end of the file, you basically pretend the file is padded with 0 until the final size.

This is harmless since it will just make a part of the tree invalid, which is perfectly fine and expected for a partial outboard.